web/src/classic/core/mantle/data/shapefile/parseZip.ts

```diff
@@ -0,0 +1,89 @@
+import JSZip from "jszip";
+import proj4 from "proj4";
+
+import { parseDbf } from "./parseDbf";
+import { parseShp } from "./parseShp";
+import { combine } from "./shapefile";
+
+function toBuffer(b: ArrayBuffer | Buffer): Buffer {
+  if (!b) {
+    throw new Error("forgot to pass buffer");
+  }
+  if (Buffer.isBuffer(b)) {
+    return b;
+  }
+  if (b instanceof ArrayBuffer) {
+    return Buffer.from(b);
+  }
+  throw new Error("Unknown buffer type");
+}
+
+export async function parseZip(
+  buffer: ArrayBuffer | Buffer,
+  whiteList?: string[],
+): Promise<GeoJSON.GeoJSON | GeoJSON.GeoJSON[]> {
+  buffer = toBuffer(buffer);
+  const zip = await JSZip.loadAsync(buffer);
+  const names: string[] = [];
+  const projections: { [key: string]: proj4.Converter } = {};
+  whiteList = whiteList || [];
+
+  for (const key of Object.keys(zip.files)) {
+    if (key.indexOf("__MACOSX") !== -1) {
+      continue;
+    }
+
+    const fileExt = key.slice(-4).toLowerCase();
+    const fileName = key.slice(0, -4);
+
+    if (fileExt === ".shp") {
+      names.push(fileName);
+      zip.files[fileName + ".shp"] = zip.files[key];
+    } else if (fileExt === ".prj") {
+      projections[fileName] = proj4(await zip.files[key].async("text"));
+    } else if (
+      fileExt === ".json" ||
+      (whiteList && whiteList.indexOf(key.split(".").pop() || "") > -1)
+    ) {
+      names.push(fileName);
+    } else if (fileExt === ".dbf" || fileExt === ".cpg") {
+      zip.files[fileName + fileExt] = zip.files[key];
+    }
+  }
+
+  if (!names.length) {
+    throw new Error("no layers found");
+  }
+
+  const geojson = await Promise.all(
+    names.map(async name => {
+      const lastDotIdx = name.lastIndexOf(".");
+      if (lastDotIdx > -1 && name.slice(lastDotIdx).indexOf("json") > -1) {
+        const parsed = JSON.parse(await zip.files[name].async("text"));
+        parsed.fileName = name.slice(0, lastDotIdx);
+        return parsed;
+      } else if (whiteList && whiteList.indexOf(name.slice(lastDotIdx + 1)) > -1) {
```

Comment:
```suggestion
      } else if (whiteList.indexOf(name.slice(lastDotIdx + 1)) > -1) {
```
Seems `whiteList` is always defined.

Reply:
BTW isn't `whiteList` used?

Reply:
whiteList isn't always defined, it was optional parameter but I've removed as we aren't really using it...